File: rust/500.keyboard-row.rs
/*
 * @lc app=leetcode id=500 lang=rust
 *
 * [500] Keyboard Row
 */

// @lc code=start
impl Solution {
    pub fn find_words(words: Vec<String>) -> Vec<String> {
        let keyboard = [
            2, 3, 3, 2, 1, 2, 2, 2, 1, 2, 2, 2, 3, 3, 1, 1, 1, 1, 2, 1, 1, 3, 1, 3, 1, 3,
        ];
        words
            .into_iter()
            .filter(|word| {
                let lowercase_word = word.to_ascii_lowercase();
                let mut bytes = lowercase_word.bytes();
                let r = keyboard[(bytes.next().unwrap() - b'a') as usize];
                bytes.all(|c| keyboard[(c - b'a') as usize] == r)
            })
            .collect()
    }
}
// @lc code=end

File: javascript/500.keyboard-row.js
/*
 * @lc app=leetcode id=500 lang=javascript
 *
 * [500] Keyboard Row
 */

// @lc code=start
/**
 * @param {string[]} words
 * @return {string[]}
 */
var findWords = function (words) {
  const keyboard = [
    2, 3, 3, 2, 1, 2, 2, 2, 1, 2, 2, 2, 3, 3, 1, 1, 1, 1, 2, 1, 1, 3, 1, 3, 1,
    3,
  ];
  const d = "a".charCodeAt();
  return words.filter((word) => {
    const lowercaseWord = word.toLowerCase();
    const r = keyboard[lowercaseWord[0].charCodeAt() - d];
    for (let i = 1; i < lowercaseWord.length; i++) {
      if (keyboard[lowercaseWord[i].charCodeAt() - d] !== r) return false;
    }
    return true;
  });
};
// @lc code=end
